pkg/externalresource/manager/service.go

```diff
@@ -54,5 +85,121 @@ func (s *Service) CreateResource(
 		return nil, st.Err()
 	}
 
+	// Note that the cache is updated AFTER the metastore is updated.
+	if _, exists := s.cache[request.GetResourceId()]; exists {
+		log.L().Panic("Cache is inconsistent",
+			zap.String("resource-id", request.GetResourceId()))
+	}
+	s.cache[request.GetResourceId()] = resourceRecord
+
 	return &pb.CreateResourceResponse{}, nil
 }
+
+func (s *Service) GetPlacementConstraint(
+	ctx context.Context,
+	id resourcemeta.ResourceID,
+) (resourcemeta.ExecutorID, bool, error) {
+	logger := log.L().WithFields(zap.String("resource-id", id))
+
+	rType, _, err := resourcemeta.ParseResourcePath(id)
+	if err != nil {
+		return "", false, err
+	}
+
+	if rType != resourcemeta.ResourceTypeLocalFile {
+		logger.Info("Resource does not need a constraint",
+			zap.String("resource-id", id), zap.String("type", string(rType)))
+		return "", false, nil
+	}
+
+	if !s.mu.Lock(ctx) {
+		return "", false, errors.Trace(ctx.Err())
+	}
+	defer s.mu.Unlock()
+
+	record, exists := s.cache[id]
+	if !exists {
+		logger.Info("Resource cache miss")
+		var err error
+
+		startTime := time.Now()
+		record, exists, err = s.accessor.GetResource(ctx, id)
+		getResourceDuration := time.Since(startTime)
+
+		logger.Info("Resource meta fetch completed", zap.Duration("duration", getResourceDuration))
+
+		if err != nil {
+			return "", false, err
+		}
+		if !exists {
+			return "", false, derror.ErrResourceDoesNotExist.GenWithStackByArgs(id)
+		}
```

Comment:
can update cache

Reply:
I'm still thinking whether it is necessary to do negative caching